stl/inc/mutex

```diff
@@ -125,7 +125,7 @@ _INLINE_VAR constexpr try_to_lock_t try_to_lock{};
 
 // CLASS TEMPLATE unique_lock
 template <class _Mutex>
-class unique_lock { // whizzy class with destructor that unlocks mutex
+class _NODISCARD_CTOR unique_lock { // whizzy class with destructor that unlocks mutex

```

Comment:
Why are we using `_NODISCARD_CTOR` on type definitions instead of `_NODISCARD`? Applying `[[nodiscard]]` to the definition of a class or enumeration was part of the original feature. (Several occurrences.)

Reply:
I agree that it would be fine to change. Checking old Clang with Compiler Explorer (Clang 8), `[[nodiscard]]` was accepted here and emitted warnings for `MyClass{}` in `-std=c++14` and `-std=c++17` mode (although, curiously, not in `-std=c++2a` mode). With modern MSVC and Clang, it's essentially moot because the feature-test macro is always defined (although Clang varies its diagnostic with the Standard version):

```
C:\Temp>type meow.cpp
#include <cstdio>
using namespace std;

struct [[nodiscard]] UniqueLock {
    UniqueLock() = default;
};

int main() {
#if __has_cpp_attribute(nodiscard) >= 201907L
    puts("__has_cpp_attribute(nodiscard) >= 201907L");
#else
    puts("__has_cpp_attribute(nodiscard) < 201907L");
#endif
    UniqueLock{};
}

C:\Temp>cl /EHsc /nologo /W4 /std:c++14 meow.cpp && meow
meow.cpp
meow.cpp(14): warning C4834: discarding return value of function with 'nodiscard' attribute
__has_cpp_attribute(nodiscard) >= 201907L

C:\Temp>cl /EHsc /nologo /W4 /std:c++latest meow.cpp && meow
meow.cpp
meow.cpp(14): warning C4834: discarding return value of function with 'nodiscard' attribute
__has_cpp_attribute(nodiscard) >= 201907L

C:\Temp>clang-cl /EHsc /nologo /W4 /std:c++14 meow.cpp && meow
meow.cpp(14,5): warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
    UniqueLock{};
    ^         ~~
1 warning generated.
__has_cpp_attribute(nodiscard) >= 201907L

C:\Temp>clang-cl /EHsc /nologo /W4 /std:c++latest meow.cpp && meow
meow.cpp(14,5): warning: ignoring temporary created by a constructor declared with 'nodiscard' attribute
      [-Wunused-value]
    UniqueLock{};
    ^~~~~~~~~~~~
1 warning generated.
__has_cpp_attribute(nodiscard) >= 201907L
```

EDG is also fine with this, and I believe CUDA will be too.